chainermn/datasets/shuffle_datablock.py

```diff
@@ -0,0 +1,219 @@
+            send_counts[(rank, rank)] = \
+                block_lengths[rank] - diff_length[rank]
+        else:
+            # Otherwise, the rank sents not data and
+            # just keeps the block for itself
+            send_counts[(rank, rank)] = block_lengths[rank]
+
+    # calculate the all-to-all send_counts as a dict.
+    # NOTE: mpi4py's alltoall() is similar to MPI_Alltoallv().
+    #       mpi4py does not have alltoallv().
+    # dst_rank and src_rank go through diff_length array once so O(N)
+    dst_rank = 0
+    for src_rank in range(size):
+        while diff_length[src_rank] > 0:
+            # src_rank has elements to send to other ranks
+            # find receiver(s)
+            while diff_length[dst_rank] >= 0:
+                dst_rank = (dst_rank + 1) % size
+            send_cnt = diff_length[src_rank]
+            recv_cnt = diff_length[dst_rank]
+            if send_cnt <= -recv_cnt:
+                # dst_rank can recieve more elements than
+                # src_rank can send. Thus src_rank's send_cnt becomes 0
+                send_counts[(src_rank, dst_rank)] = send_cnt
+                diff_length[src_rank] = 0
+                diff_length[dst_rank] += send_cnt
+            else:
+                # dst_rank can receive less elements than
+                # src_rank can send. Thus dst_rank's recv_cnt becomes 0
+                # NOTE: recv_cnt is a <0 value
+                send_counts[(src_rank, dst_rank)] = -recv_cnt
+                diff_length[src_rank] += recv_cnt
+                diff_length[dst_rank] = 0
+            if diff_length[src_rank] == 0:
+                break
+
+    # All the diffs are resolved by communication.
+    assert all(d == 0 for d in diff_length)
+    return send_counts, new_lengths
+
+
+def _exchange_block(comm, data, block, data_lengths, block_lengths):
+    send_counts, new_lengths = _calc_alltoall_send_counts(
+        data_lengths, block_lengths)
+
+    # Basically, send data is selected from the newly-loaded `block`,
+    # but in some cases `block` is an empty array
+    # i.e. when adjusting for force_equal_length
+    if len(block) == 0:
+        local_data = data
+    else:
+        local_data = block
+
+    offset = 0
+    send_buf = [[]] * comm.size
+    for dest_rank in range(comm.size):
+        num_elem = send_counts.get((comm.rank, dest_rank), 0)
+        if num_elem == 0:
+            continue
+
+        send_buf[dest_rank] = local_data[offset:offset + num_elem]
+        while len(send_buf[dest_rank]) < num_elem:
+            # In case of force_equal_length, the process may have to send
+            # more data than `block` has. We need to duplicate some elements
+            # from `block`.
+            send_buf[dest_rank].append(random.choice(block))
+        offset = (offset + num_elem) % len(local_data)
+    data += _flatten1(comm.mpi_comm.alltoall(send_buf))
+    return new_lengths
+
+
+def shuffle_data_blocks(comm, blocks, block_size, force_equal_length=True):
```

Comment:
IIRC the definition of "block" has been changed in this change since we discussed offline. The original use case was that each process reads arbitrary number of .npy files that includes arbitrary number of records, where each .npy file was called "block". That said, the definition of block is a set of data element (e.g. example in dataset mixin). And the argument of the function was an iterator that iterates over them.

Otherwise, the term "block" is can be changed to more intuitive word like "elements", "examples" and so on.